File: 算法/代码随想录算法PDF/00调试源代码/110.平衡二叉树.cpp
/*
 * @lc app=leetcode.cn id=110 lang=cpp
 *
 * [110] 平衡二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (!root) return true;
        stack<TreeNode*> st;
        unordered_map<TreeNode*, int> map;
        st.push(root);
        int left, right;  // 左右子节点高度
        while (!st.empty()) {
            TreeNode* node = st.top();
            if (node) {  // 不为空，说明是第一次遍历
                st.push(nullptr);
                if (node->right) st.push(node->right);
                if (node->left) st.push(node->left);
            } else {
                st.pop();
                node = st.top();
                st.pop();
                if (map.find(node->left) == map.end()) {  // map中没找到，默认为0
                    left = 0;
                } else {
                    left = map[node->left];
                }
                if (map.find(node->right) == map.end()) {
                    right = 0;
                } else {
                    right = map[node->right];
                }

                if (abs(left - right) > 1) return false;
                map[node] = 1 + max(left, right);
            }
        }
        return true;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/110.平衡二叉树.py
#
# @lc app=leetcode.cn id=110 lang=python3
#
# [110] 平衡二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        st = [root]
        map = {}
        while st:
            node = st.pop()
            if node:
                st.append(node)
                st.append(None)
                if node.right: st.append(node.right)
                if node.left: st.append(node.left)
            else:
                node = st.pop()
                left, right = map.get(node.left, 0), map.get(node.right, 0)
                if abs(left - right) > 1:
                    return False
                map[node] = 1 + max(left, right)
        return True
# @lc code=end

